File: src/jzrlib/map/dungeons/CustomStaticArea.java
/*
 *  "NorseWorld: Ragnarok", a roguelike game for PCs.
 *  Copyright (C) 2002-2008, 2014 by Serg V. Zhdanovskih (aka Alchemist).
 *
 *  This file is part of "NorseWorld: Ragnarok".
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package jzrlib.map.dungeons;

import java.util.ArrayList;
import jzrlib.core.Directions;
import jzrlib.map.TileType;
import jzrlib.utils.AuxUtils;
import jzrlib.utils.Logger;

/**
 * 
 * @author Serg V. Zhdanovskih (aka Alchemist), 2005, 2014-2015
 */
public abstract class CustomStaticArea extends StaticArea
{
    private static final class CustomMark
    {
        public int x;
        public int y;
        public int dir;

        public CustomMark(int x, int y, int dir)
        {
            this.x = x;
            this.y = y;
            this.dir = dir;
        }
    }

    private static final String DoorSym = "NSWE";
    private static final String CustomSym = "X.NSWE";

    private ArrayList<CustomMark> fAvailMarks;
    protected int fDevourArea;

    public CustomStaticArea(DungeonBuilder owner, DungeonMark parentMark)
    {
        super(owner, parentMark);
        this.fAvailMarks = new ArrayList<>();
    }

    @Override
    protected void dispose(boolean disposing)
    {
        if (disposing) {
            this.fAvailMarks.clear();
            this.fAvailMarks = null;
        }
        super.dispose(disposing);
    }

    private void addDoor(ArrayList<CustomMark> doorsList, int aX, int aY, char aSym)
    {
        int dir;

        switch (aSym) {
            case 'E':
                dir = Directions.dtEast;
                break;
            case 'N':
                dir = Directions.dtNorth;
                break;
            case 'S':
                dir = Directions.dtSouth;
                break;
            case 'W':
                dir = Directions.dtWest;
                break;
            default:
                return;
        }

        doorsList.add(new CustomMark(aX, aY, dir));
    }

    @Override
    protected boolean buildArea()
    {
        boolean result = false;

        try {
            this.fAvailMarks.clear();
            this.fDevourArea = 0;

            ArrayList<CustomMark> doors = new ArrayList<>();

            for (int y = 0; y < super.Height; y++) {
                for (int x = 0; x < super.Width; x++) {
                    char sym = this.getCustomCell(x, y);
                    if (sym == '.') {
                        this.fDevourArea++;
                        super.fArea[x][y] = TileType.ttUndefined;
                    } else {
                        if (sym == 'E' || sym == 'N' || sym == 'S' || sym == 'W' || sym == 'X') {
                            this.fDevourArea++;
                            super.fArea[x][y] = TileType.ttDungeonWall;

                            if (CustomStaticArea.DoorSym.indexOf(sym) >= 0) {
                                this.addDoor(fAvailMarks, x, y, sym);
                                if (sym == Directions.Data[super.ParentMark.Direction].symOpposite) {
                                    this.addDoor(doors, x, y, sym);
                                }
                            }
                        }
                    }
                }
            }

            CustomMark door = AuxUtils.getRandomItem(doors);
            if (door != null) {
                super.setPosition(super.ParentMark.Location.X - door.x, super.ParentMark.Location.Y - door.y);
                result = true;
            }
        } catch (Exception ex) {
            Logger.write("CustomStaticArea.buildArea(): " + ex.getMessage());
            throw ex;
        }

        return result;
    }

    protected abstract char getCustomCell(int aX, int aY);

    @Override
    public void generateMarksList()
    {
        try {
            for (CustomMark mark : this.fAvailMarks) {
                super.tryInsertMark(this.Left + mark.x, this.Top + mark.y, mark.dir);
            }
        } catch (Exception ex) {
            Logger.write("CustomStaticArea.generateMarksList(): " + ex.getMessage());
            throw ex;
        }
    }

    @Override
    public int getDevourArea()
    {
        return this.fDevourArea;
    }

    @Override
    public boolean isOwnedPoint(int ptX, int ptY)
    {
        int ax = ptX - this.Left;
        int ay = ptY - this.Top;

        boolean result;
        if (ax < 0 || ax >= super.Width || ay < 0 || ay >= super.Height) {
            result = false;
        } else {
            char sym = this.getCustomCell(ax, ay);
            result = CustomStaticArea.CustomSym.indexOf(sym) >= 0;
        }
        return result;
    }
}

File: project/Map/Dungeons/CustomStaticArea.cs
/*
 *  "NorseWorld: Ragnarok", a roguelike game for PCs.
 *  Copyright (C) 2002-2008, 2014 by Serg V. Zhdanovskih.
 *
 *  This file is part of "NorseWorld: Ragnarok".
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

using System;
using System.Collections.Generic;
using BSLib;
using ZRLib.Core;

namespace ZRLib.Map.Dungeons
{
    public abstract class CustomStaticArea : StaticArea
    {
        private sealed class CustomMark
        {
            public int X;
            public int Y;
            public int Dir;

            public CustomMark(int x, int y, int dir)
            {
                X = x;
                Y = y;
                Dir = dir;
            }
        }

        private const string DoorSym = "NSWE";
        private const string CustomSym = "X.NSWE";

        private List<CustomMark> fAvailMarks;
        protected int fDevourArea;

        protected CustomStaticArea(DungeonBuilder owner, DungeonMark parentMark)
            : base(owner, parentMark)
        {
            fAvailMarks = new List<CustomMark>();
        }

        protected override void Dispose(bool disposing)
        {
            if (disposing) {
                fAvailMarks.Clear();
                fAvailMarks = null;
            }
            base.Dispose(disposing);
        }

        private static void AddDoor(List<CustomMark> doorsList, int aX, int aY, char aSym)
        {
            int dir;

            switch (aSym) {
                case 'E':
                    dir = Directions.DtEast;
                    break;
                case 'N':
                    dir = Directions.DtNorth;
                    break;
                case 'S':
                    dir = Directions.DtSouth;
                    break;
                case 'W':
                    dir = Directions.DtWest;
                    break;
                default:
                    return;
            }

            doorsList.Add(new CustomMark(aX, aY, dir));
        }

        protected override bool BuildArea()
        {
            bool result = false;

            try {
                fAvailMarks.Clear();
                fDevourArea = 0;

                List<CustomMark> doors = new List<CustomMark>();

                for (int y = 0; y < Height; y++) {
                    for (int x = 0; x < Width; x++) {
                        char sym = GetCustomCell(x, y);
                        if (sym == '.') {
                            fDevourArea++;
                            fArea[x, y] = TileType.ttUndefined;
                        } else {
                            if (sym == 'E' || sym == 'N' || sym == 'S' || sym == 'W' || sym == 'X') {
                                fDevourArea++;
                                fArea[x, y] = TileType.ttDungeonWall;

                                if (DoorSym.IndexOf(sym) >= 0) {
                                    AddDoor(fAvailMarks, x, y, sym);
                                    if (sym == Directions.Data[ParentMark.Direction].SymOpposite) {
                                        AddDoor(doors, x, y, sym);
                                    }
                                }
                            }
                        }
                    }
                }

                CustomMark door = RandomHelper.GetRandomItem(doors);
                if (door != null) {
                    SetPosition(ParentMark.Location.X - door.X, ParentMark.Location.Y - door.Y);
                    result = true;
                }
            } catch (Exception ex) {
                Logger.Write("CustomStaticArea.buildArea(): " + ex.Message);
                throw ex;
            }

            return result;
        }

        protected abstract char GetCustomCell(int aX, int aY);

        public override void GenerateMarksList()
        {
            try {
                foreach (CustomMark mark in fAvailMarks) {
                    TryInsertMark(Left + mark.X, Top + mark.Y, mark.Dir);
                }
            } catch (Exception ex) {
                Logger.Write("CustomStaticArea.generateMarksList(): " + ex.Message);
                throw ex;
            }
        }

        public override int DevourArea
        {
            get { return fDevourArea; }
        }

        public override bool IsOwnedPoint(int ptX, int ptY)
        {
            int ax = ptX - Left;
            int ay = ptY - Top;

            bool result;
            if (ax < 0 || ax >= Width || ay < 0 || ay >= Height) {
                result = false;
            } else {
                char sym = GetCustomCell(ax, ay);
                result = CustomSym.IndexOf(sym) >= 0;
            }
            return result;
        }
    }
}
